modules/xfem/src/userobjects/InterfaceMeshCut3DUserObject.C

```diff
@@ -0,0 +1,406 @@
+
+      _explicit_system->get_dof_map().dof_indices(
+          node, di, _explicit_system->variable_number("disp_z"));
+      _explicit_system->solution->set(
+          di[0], new_position[node->id()](2) - _initial_nodes_location[node->id()](2));
+    }
+
+    _explicit_system->solution->close();
+
+    _exodus_io->append(true);
+    _exodus_io->write_timestep(
+        _app.getOutputFileBase() + "_" + name() + ".e", *_equation_systems, _t_step + 1, _t);
+  }
+
+  _pseudo_normal.clear();
+
+  for (const auto & elem : _cutter_mesh->element_ptr_range())
+  {
+    std::vector<Point> vertices{elem->node_ref(0), elem->node_ref(1), elem->node_ref(2)};
+    std::array<Point, 7> normal;
+    Plane elem_plane(vertices[0], vertices[1], vertices[2]);
+    normal[0] = 2.0 * libMesh::pi * elem_plane.unit_normal(vertices[0]);
+
+    for (unsigned int i = 0; i < elem->n_nodes(); i++)
+    {
+      Point normal_at_node(0.0);
+      const Node & node = elem->node_ref(i);
+
+      Real angle_sum = 0.0;
+
+      for (const auto & node_neigh_elem_id : _node_to_elem_map[node.id()])
+      {
+        const Elem & node_neigh_elem = _cutter_mesh->elem_ref(node_neigh_elem_id);
+        std::vector<Point> vertices{
+            node_neigh_elem.node_ref(0), node_neigh_elem.node_ref(1), node_neigh_elem.node_ref(2)};
+        Plane elem_plane(vertices[0], vertices[1], vertices[2]);
+        unsigned int j = node_neigh_elem.local_node(node.id());
+        Point normal_at_node_j = elem_plane.unit_normal(vertices[0]);
+        unsigned int m = j + 1 < 3 ? j + 1 : j + 1 - 3;
+        unsigned int n = j + 2 < 3 ? j + 2 : j + 2 - 3;
+        Point line_1 = node_neigh_elem.node_ref(j) - node_neigh_elem.node_ref(m);
+        Point line_2 = node_neigh_elem.node_ref(j) - node_neigh_elem.node_ref(n);
+        Real dot = line_1 * line_2;
+        Real lenSq1 = line_1 * line_1;
+        Real lenSq2 = line_2 * line_2;
+        Real angle = std::acos(dot / std::sqrt(lenSq1 * lenSq2));
+        normal_at_node += normal_at_node_j * angle;
+        angle_sum += angle;
+      }
+      normal[1 + i] = normal_at_node;
+    }
+
+    for (unsigned int i = 0; i < elem->n_sides(); i++)
+    {
+      std::vector<Point> vertices{elem->node_ref(0), elem->node_ref(1), elem->node_ref(2)};
+
+      Plane elem_plane(vertices[0], vertices[1], vertices[2]);
+      Point normal_at_edge = libMesh::pi * elem_plane.unit_normal(vertices[0]);
+
+      const Elem * neighbor = elem->neighbor_ptr(i);
+
+      if (neighbor != nullptr)
+      {
+        std::vector<Point> vertices{
+            neighbor->node_ref(0), neighbor->node_ref(1), neighbor->node_ref(2)};
+
+        Plane elem_plane(vertices[0], vertices[1], vertices[2]);
+        normal_at_edge += libMesh::pi * elem_plane.unit_normal(vertices[0]);
+      }
+      normal[4 + i] = normal_at_edge;
+    }
+    _pseudo_normal.insert(std::make_pair(elem->id(), normal));
+  }
+}
+
+Point
+InterfaceMeshCut3DUserObject::nodeNomal(const unsigned int & node_id)
```

Comment:
```suggestion
InterfaceMeshCut3DUserObject::nodeNormal(const unsigned int & node_id)
```

I think these nodeNormal methods are useful in general -- not just in InterfaceMeshCut userobjects. It is better to sit in XFEM/utils. We also need some unit tests for it. The function signature can be `nodeNomal(const std::vector<const Elem *> elements)`, then inside the function, we can use a switch statement to dispatch it to helper functions like `nodeNormalTRI3`, `nodeNormalEDGE2`, etc.. If it is not implemented, we can throw an error there, so that you don't need to add an element type check in the constructor of InterfaceMeshCut userobjects. It's also easier to extend in the future.

Reply:
where else do you think we need nodeNormal function?